packages/core/src/transport/sendWithRetryStrategy.ts

```diff
@@ -0,0 +1,153 @@
+export const MAX_ONGOING_REQUESTS = 32
+export const MAX_QUEUE_BYTES_COUNT = 3 * ONE_MEGA_BYTE
+export const MAX_BACKOFF_TIME = 256 * ONE_SECOND
+
+export interface RetryState {
+  isIntakeAvailable: boolean
+  currentBackoffTime: number
+  bandwidthMonitor: ReturnType<typeof newBandwidthMonitor>
+  queuedPayloads: ReturnType<typeof newPayloadQueue>
+}
+
+type SendStrategy = (payload: Payload, onResponse: (r: HttpResponse) => void) => void
+
+export function sendWithRetryStrategy(payload: Payload, state: RetryState, sendStrategy: SendStrategy) {
+  if (state.isIntakeAvailable && state.bandwidthMonitor.canHandle(payload)) {
+    send(payload, state, sendStrategy, {
+      onSuccess: () => sendNextPayload(state, sendStrategy),
+      onFailure: () => {
+        state.queuedPayloads.enqueue(payload)
+        scheduleRetry(state, sendStrategy)
+      },
+    })
+    sendNextPayload(state, sendStrategy)
+  } else {
+    state.queuedPayloads.enqueue(payload)
+  }
+}
+
+function scheduleRetry(state: RetryState, sendStrategy: SendStrategy) {
+  if (state.bandwidthMonitor.ongoingRequestCount !== 0) {
+    // avoid to enter a retry phase if another ongoing request can succeed
+    return
+  }
+  setTimeout(
+    monitor(() => {
+      const payload = state.queuedPayloads.first()
+      send(payload, state, sendStrategy, {
+        onSuccess: () => {
+          state.queuedPayloads.dequeue()
+          addTelemetryDebug('resuming after intake failure', {
+            currentBackoffTime: state.currentBackoffTime,
+            queuedPayloadCount: state.queuedPayloads.size(),
+            queuedPayloadBytesCount: state.queuedPayloads.bytesCount,
+          })
+          state.currentBackoffTime = ONE_SECOND
+          sendNextPayload(state, sendStrategy)
+        },
+        onFailure: () => {
+          state.currentBackoffTime = Math.min(MAX_BACKOFF_TIME, state.currentBackoffTime * 2)
+          scheduleRetry(state, sendStrategy)
+        },
+      })
+    }),
+    state.currentBackoffTime
+  )
+}
+
+function sendNextPayload(state: RetryState, sendStrategy: SendStrategy) {
+  const nextPayload = state.queuedPayloads.dequeue()
+  if (nextPayload) {
+    sendWithRetryStrategy(nextPayload, state, sendStrategy)
+  }
+}
+
+function send(
+  payload: Payload,
+  state: RetryState,
+  sendStrategy: SendStrategy,
+  { onSuccess, onFailure }: { onSuccess: () => void; onFailure: () => void }
+) {
+  state.bandwidthMonitor.add(payload)
+  sendStrategy(payload, (response) => {
+    state.bandwidthMonitor.remove(payload)
+    if (wasRequestSuccessful(response)) {
+      state.isIntakeAvailable = true
+      onSuccess()
+    } else {
+      state.isIntakeAvailable = false
+      onFailure()
+    }
+  })
+}
+
+function wasRequestSuccessful(response: HttpResponse) {
+  return response.status < 500
+}
+
+export function newRetryState(): RetryState {
+  return {
+    isIntakeAvailable: true,
+    currentBackoffTime: ONE_SECOND,
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏use a constant like `INITIAL_BACKOFF_TIME`